modules/redis/redis.c

```diff
@@ -290,15 +294,82 @@ redis_worker_insert(LogThreadedDestDriver *s, LogMessage *msg)
       return LTR_ERROR;
     }
 
-  msg_debug("REDIS command sent",
-            evt_tag_str("driver", self->super.super.super.id),
-            evt_tag_str("command", full_command->str));
+  if (debug_flag)
+    {
+      GString *full_command = _argv_to_string(self);
+      msg_debug("REDIS command sent",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("command", full_command->str));
+    }
 
   freeReplyObject(reply);
   scratch_buffers_reclaim_marked(marker);
   return LTR_SUCCESS;
 }
 
+static LogThreadedResult
+redis_worker_insert_batch_mode(LogThreadedDestDriver *s, LogMessage *msg)
+{
+  RedisDriver *self = (RedisDriver *)s;
+
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+
+  _fill_argv_from_template_list(self, msg);
+
+  gint retval = redisAppendCommandArgv(self->c, self->argc, (const gchar **)self->argv, self->argvlen);
+  if (retval != REDIS_OK)
+    return LTR_ERROR;
+
+  if (debug_flag)
+    {
+      GString *full_command = _argv_to_string(self);
+      msg_debug("REDIS command pipelined",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("command", full_command->str));
+    }
+
+  return LTR_QUEUED;
+}
+
+static LogThreadedResult
+redis_worker_flush(LogThreadedDestDriver *s)
+{
+  RedisDriver *self = (RedisDriver *)s;
+
+  if (s->worker.instance.batch_size == 0)
+    return LTR_SUCCESS;
+
+  LogThreadedResult connection_status = redis_dd_connect(self);
+  if (connection_status != LTR_SUCCESS)
+    return connection_status;
+
+  void *reply;
+  if (redisGetReply(self->c, &reply) != REDIS_OK)
+    {
+      msg_error("REDIS server error during pipelined send");
+      return LTR_ERROR;
+    }
+
+  freeReplyObject(reply);
+
+  msg_debug("REDIS pipelined command sent",
+            evt_tag_str("driver", self->super.super.super.id));
+
+  return LTR_SUCCESS;
+}
+
+static LogThreadedResult
+redis_worker_insert_selector(LogThreadedDestDriver *s, LogMessage *msg)
+{
+  if (s->batch_lines > 1)
+    s->worker.insert = redis_worker_insert_batch_mode;
+  else
+    s->worker.insert = redis_worker_insert;
+
+  return s->worker.insert(s, msg);
```

Comment:
This might be executing in multiple threads, which probably wouldnt cause problems, but I would do this initialization during init()

Reply:
Moved into the init phase, thank you. Also managed to get rid of the "selector" function. By defaulting the non batching mode.